### PR TITLE
[pipelineX](refactor) propose a new pipeline execution model

### PR DESCRIPTION
**Pull request by Gabriel39**
## Proposed change

Propose a new pipeline execution model. Different from the former one, we use a shared operator tree and all global state for each pipeline task and separated local states. This will avoid the unnecessary overhead for initializing and reduce the RPC overhead between BE instances.

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by Gabriel39** [comment]
run buildall

**Comment by Gabriel39** [comment]
run buildall

**Comment by Gabriel39** [comment]
run buildall

**Comment by Gabriel39** [comment]
run buildall

**Comment by Gabriel39** [comment]
run buildall

**Comment by Gabriel39** [comment]
run buildall

**Comment by Gabriel39** [comment]
run buildall

**Comment by Gabriel39** [comment]
run buildall

**Comment by Gabriel39** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 45.24 seconds
 stream load tsv:          544 seconds loaded 74807831229 Bytes, about 131 MB/s
 stream load json:         21 seconds loaded 2358488459 Bytes, about 107 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          32 seconds loaded 861443392 Bytes, about 25 MB/s
 insert into select:          29.2 seconds inserted 10000000 Rows, about 342K ops/s
 storage size: 17162218403 Bytes

**Review by BiteTheDDDDt — APPROVED** [review]

**Review by yiguolei — APPROVED** [review]
LGTM

File: be/src/pipeline/pipeline_x/pipeline_x_fragment_context.cpp [hunk 1]
```diff
@@ -0,0 +1,585 @@
...
+#include "vec/exec/scan/vscan_node.h"
+#include "vec/exec/vaggregation_node.h"
+#include "vec/exec/vexchange_node.h"
+#include "vec/exec/vunion_node.h"
+#include "vec/runtime/vdata_stream_mgr.h"
+
+namespace doris::pipeline {
+
+#define FOR_EACH_RUNTIME_STATE(stmt)              \
+    for (auto& runtime_state : _runtime_states) { \
```
Comment (BiteTheDDDDt) [inline be/src/pipeline/pipeline_x/pipeline_x_fragment_context.cpp:81]: Try not to use macros. Using macros in these places doesn't simplify the code much.

File: be/src/pipeline/pipeline_x/pipeline_x_fragment_context.cpp [hunk 2]
```diff
@@ -0,0 +1,585 @@
...
+        if (!thrift_sink.__isset.stream_sink) {
+            return Status::InternalError("Missing data stream sink.");
+        }
+        bool send_query_statistics_with_every_batch =
+                params.__isset.send_query_statistics_with_every_batch
+                        ? params.send_query_statistics_with_every_batch
+                        : false;
+        _sink.reset(new ExchangeSinkOperatorX(
+                _sink_idx++, state, pool, row_desc, thrift_sink.stream_sink, params.destinations,
+                16 * 1024, send_query_statistics_with_every_batch, this));
```
Comment (BiteTheDDDDt) [inline be/src/pipeline/pipeline_x/pipeline_x_fragment_context.cpp:241]: Set a constant and explain the meaning of 16 * 1024

File: be/src/pipeline/pipeline_x/pipeline_x_fragment_context.cpp [hunk 3]
```diff
@@ -0,0 +1,585 @@
...
+        break;
+    }
+    case TDataSinkType::RESULT_SINK: {
+        if (!thrift_sink.__isset.result_sink) {
+            return Status::InternalError("Missing data buffer sink.");
+        }
+
+        // TODO: figure out good buffer size based on size of output row
+        _sink.reset(new ResultSinkOperatorX(_sink_idx++, row_desc, output_exprs,
+                                            thrift_sink.result_sink, 4096));
```
Comment (BiteTheDDDDt) [inline be/src/pipeline/pipeline_x/pipeline_x_fragment_context.cpp:251]: ditto of 4096

File: be/src/pipeline/pipeline_x/pipeline_x_fragment_context.cpp [hunk 4]
```diff
@@ -0,0 +1,585 @@
...
+
+            std::unique_lock<std::mutex> l(_report_thread_lock);
+            // tell Open() that we started
+            _report_thread_started_cv.notify_one();
+
+            // Jitter the reporting time of remote fragments by a random amount between
+            // 0 and the report_interval.  This way, the coordinator doesn't get all the
+            // updates at once so its better for contention as well as smoother progress
+            // reporting.
+            int report_fragment_offset = rand() % config::status_report_interval;
```
Comment (BiteTheDDDDt) [inline be/src/pipeline/pipeline_x/pipeline_x_fragment_context.cpp:334]: maybe status_report_interval is useless

File: be/src/pipeline/pipeline_x/pipeline_x_fragment_context.cpp [hunk 5]
```diff
@@ -0,0 +1,585 @@
...
+                            l, std::chrono::seconds(config::status_report_interval));
+                } else {
+                    LOG(WARNING) << "config::status_report_interval is equal to or less than zero, "
+                                    "exiting "
+                                    "reporting thread.";
+                    break;
+                }
+
+                if (VLOG_FILE_IS_ON) {
+                    VLOG_FILE << "Reporting " << (!_report_thread_active ? "final " : " ")
```
Comment (BiteTheDDDDt) [inline be/src/pipeline/pipeline_x/pipeline_x_fragment_context.cpp:354]: Maybe these vlogs can all be removed, I'm not sure

File: be/src/pipeline/pipeline_x/pipeline_x_fragment_context.cpp [hunk 6]
```diff
@@ -0,0 +1,585 @@
...
+            VLOG_FILE
+            << "exiting reporting thread: instance_id=" << runtime_state->fragment_instance_id();)
+}
+
+Status PipelineXFragmentContext::_build_pipelines(ObjectPool* pool,
+                                                  const doris::TPipelineFragmentParams& request,
+                                                  const DescriptorTbl& descs, OperatorXPtr* root,
+                                                  PipelinePtr cur_pipe) {
+    if (request.fragment.plan.nodes.size() == 0) {
+        *root = nullptr;
```
Comment (BiteTheDDDDt) [inline be/src/pipeline/pipeline_x/pipeline_x_fragment_context.cpp:382]: When will size be 0?

File: be/src/pipeline/pipeline_x/pipeline_x_fragment_context.cpp [hunk 7]
```diff
@@ -0,0 +1,585 @@
...
+Status PipelineXFragmentContext::_create_tree_helper(ObjectPool* pool,
+                                                     const std::vector<TPlanNode>& tnodes,
+                                                     const doris::TPipelineFragmentParams& request,
+                                                     const DescriptorTbl& descs,
+                                                     OperatorXPtr parent, int* node_idx,
+                                                     OperatorXPtr* root, PipelinePtr& cur_pipe) {
+    // propagate error case
+    if (*node_idx >= tnodes.size()) {
+        // TODO: print thrift msg
+        return Status::InternalError("Failed to reconstruct plan tree from thrift.");
```
Comment (BiteTheDDDDt) [inline be/src/pipeline/pipeline_x/pipeline_x_fragment_context.cpp:408]: add invalid index and size into status's msg

File: be/src/pipeline/pipeline_x/pipeline_x_fragment_context.cpp [hunk 8]
```diff
@@ -0,0 +1,585 @@
...
+    for (int i = 0; i < num_children; i++) {
+        ++*node_idx;
+        RETURN_IF_ERROR(
+                _create_tree_helper(pool, tnodes, request, descs, op, node_idx, nullptr, cur_pipe));
+
+        // we are expecting a child, but have used all nodes
+        // this means we have been given a bad tree and must fail
+        if (*node_idx >= tnodes.size()) {
+            // TODO: print thrift msg
+            return Status::InternalError("Failed to reconstruct plan tree from thrift.");
```
Comment (BiteTheDDDDt) [inline be/src/pipeline/pipeline_x/pipeline_x_fragment_context.cpp:432]: ditto

File: be/src/pipeline/pipeline_x/pipeline_x_fragment_context.cpp [hunk 9]
```diff
@@ -0,0 +1,585 @@
...
+        op->get_runtime_profile()->add_child(op->get_child()->get_runtime_profile(), true, nullptr);
+    }
+
+    return Status::OK();
+}
+
+Status PipelineXFragmentContext::_create_operator(ObjectPool* pool, const TPlanNode& tnode,
+                                                  const doris::TPipelineFragmentParams& request,
+                                                  const DescriptorTbl& descs, OperatorXPtr& op,
+                                                  PipelinePtr& cur_pipe) {
```
Comment (BiteTheDDDDt) [inline be/src/pipeline/pipeline_x/pipeline_x_fragment_context.cpp:448]: Need to pass a reference to shared_ptr?

File: be/src/pipeline/pipeline_x/pipeline_x_fragment_context.cpp [hunk 10]
```diff
@@ -0,0 +1,585 @@
...
+    Status st;
+    auto* scheduler = _exec_env->pipeline_task_scheduler();
+    if (_task_group_entity) {
+        scheduler = _exec_env->pipeline_task_group_scheduler();
+    }
+    for (auto& task : _tasks) {
+        for (auto& t : task) {
+            st = scheduler->schedule_task(t.get());
+            if (!st) {
+                std::lock_guard<std::mutex> l(_status_lock);
```
Comment (BiteTheDDDDt) [inline be/src/pipeline/pipeline_x/pipeline_x_fragment_context.cpp:500]: It seems that _status_lock is also locked in PipelineXFragmentContext::cancel

**Pull request merged**